irohad/ametsuchi/impl/storage_impl.cpp

```diff
@@ -61,6 +58,102 @@ namespace {
 
 namespace iroha {
   namespace ametsuchi {
+
+    /**
+     * Class provides reconnection callback for postgresql session
+     * Note: the class is a workaround for SOCI 4.0, support in future versions
+     * is not guaranteed
+     */
+    class FailoverCallback : public soci::failover_callback {
+     public:
+      using InitFunctionType = std::function<void(soci::session &)>;
+      FailoverCallback(
+          soci::session &connection,
+          InitFunctionType init,
+          std::string connection_options,
+          std::unique_ptr<ReconnectionStrategy> reconnection_strategy,
+          logger::LoggerPtr log)
+          : connection_(connection),
+            init_function_(std::move(init)),
+            connection_options_(std::move(connection_options)),
+            reconnection_strategy_(std::move(reconnection_strategy)),
+            log_(std::move(log)) {}
+
+      FailoverCallback(const FailoverCallback &) = delete;
+      FailoverCallback &operator=(const FailoverCallback &) = delete;
+
+      virtual void started() {
+        reconnection_strategy_->reset();
+        log_->debug("Reconnection process is initiated");
+      }
+
+      virtual void finished(soci::session &) {}
+
+      virtual void failed(bool &should_reconnect, std::string &) {
+        // don't rely on reconnection in soci because we are going to conduct
+        // our own reconnection process
+        should_reconnect = false;
+        log_->warn(
+            "troubles with storage connections are occurred. The system will "
+            "try to reconnect");
+        auto is_reconnected = reconnectionLoop();
+        log_->info("re-established: {}", is_reconnected);
+      }
+
+      virtual void aborted() {
+        log_->error("has invoked aborted method of FailoverCallback");
+      }
+
+     private:
+      bool reconnectionLoop() {
+        bool successful_reconnection = false;
+        while (reconnection_strategy_->canReconnect()
+               and not successful_reconnection) {
+          try {
+            soci::connection_parameters parameters("postgresql",
```

Comment:
I am getting "Failed to find shared library for backend postgresql". Please use the following instead:
```suggestion
            soci::connection_parameters parameters(*soci::factory_postgresql(),
```